Implement Give_Cache for all transients

ravinderk opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce?  -->
<!--- Bug fix (non-breaking change which fixes an issue) -->
New feature (non-breaking change which adds functionality)
<!--- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows has proper inline documentation.

ravinderk commented:
@DevinWalker I found few transient which was deleted on certain action but I did not find where we are generating them. Please take a look on these transients.

1. https://github.com/WordImpress/Give/blob/0555aa6666d9a876928596d7ef6d8180569b7d06/includes/payments/actions.php#L170

2. https://github.com/WordImpress/Give/blob/0555aa6666d9a876928596d7ef6d8180569b7d06/includes/payments/actions.php#L88
https://github.com/WordImpress/Give/blob/0555aa6666d9a876928596d7ef6d8180569b7d06/includes/payments/actions.php#L89

3. Can you check why we are using `option` and `transient` here:
https://github.com/WordImpress/Give/blob/0555aa6666d9a876928596d7ef6d8180569b7d06/includes/admin/reporting/tools/class-give-tools-recount-income.php#L99
https://github.com/WordImpress/Give/blob/999cbdd1260bea3ce0638b5d9e335a208a0211c0/includes/payments/functions.php#L897

DevinWalker commented:
1. double check the core and all add-ons for me to see if it's being used. If not, remove it.
2. same as 1
3. Because it's easier and more of the WP way to use `get_option` when using the stats throughout your theme and custom code. We should double-check it's being updated properly. Please verify there's a unit test for this and if not add it. 

Thanks!

ravinderk commented on includes/misc-functions.php:
```diff
@@ -1050,3 +1050,35 @@ function give_is_terms_enabled( $form_id ) {
+ *
+ * @return WP_Error|bool
+ */
+function give_delete_donation_stats( $date_range = '', $args = array() ) {
```
@DevinWalker Currently, it is very difficult to get a transient key for any stat cache. That why flush all cache will be good. After refactoring reports, we can update this function too.

ravinderk commented on includes/payments/functions.php:
```diff
@@ -851,54 +852,46 @@ function give_get_total_sales() {
-function give_get_total_earnings() {
-
-	$total = get_option( 'give_earnings_total', false );
+function give_get_total_earnings( $recalculate = false ) {
```
@DevinWalker I think it will be good if we provide a param to recalculate total forcefully

ravinderk commented:
@DevinWalker 
1. Removed, because I did not find any use of this.
2. Removed, because I did not find any use of them
3. I did not find `give_earnings_total` transient useful, so I removed this from code. I also renamed`give_income_total` --> `give_earnings_total`

I successfully tested these changes by running PHPUnit tests, create new donation, refresh reports and recalculate stats.

ravinderk commented on includes/admin/reporting/tools/class-give-tools-delete-test-transactions.php:
```diff
@@ -164,10 +164,9 @@ public function process_step() {
-			delete_transient( 'give_earnings_total' );
-			delete_transient( 'give_estimated_monthly_stats' . true );
-			delete_transient( 'give_estimated_monthly_stats' . false );
+			update_option( 'give_earnings_total', give_get_total_earnings( true ) );
```
@DevinWalker I updated code to set `give_earnings_total` to latest income total instead `0`
